src/Raven.Server/Utils/CombinedReadingStream.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using System.IO;
+
+namespace Raven.Server.Utils
+{
+    public class CombinedReadingStream : Stream
+    {
+        private readonly Stream[] _streams;
+        private readonly long[] _startingPositions;
+        private long _position;
+        private readonly long _totalLength;
+        private int _index;
+
+        public CombinedReadingStream(Stream[] streams)
+        {
+            _streams = streams;
+            _startingPositions = new long[streams.Length];
+
+            _position = 0;
+            _index = 0;
+
+            _startingPositions[0] = 0;
+            for (var i = 1; i < _startingPositions.Length; i++)
+            {
+                _startingPositions[i] = _startingPositions[i - 1] + _streams[i - 1].Length;
+            }
+
+            _totalLength = _startingPositions[_startingPositions.Length - 1] + _streams[_streams.Length - 1].Length;
+        }
+
+        public override bool CanRead => true;
+
+        public override bool CanSeek => true;
+
+        public override bool CanWrite => false;
+
+        public override void Flush()
+        {
+            throw new NotSupportedException("The method or operation is not supported by CombinedReadingStream.");
+        }
+
+        public override long Length => _totalLength;
+
+        public override long Position
+        {
+            get
+            {
+                return _position;
+            }
+            set
+            {
+                if (value < 0 || value > _totalLength)
+                    throw new ArgumentOutOfRangeException("Position");
+
+                _position = value;
+
+                while (_index > 0 && _position < _startingPositions[_index])
```

Comment:
What is the value of index here? If it started with 2, the the positions of the previous streams wouldn't be set, no?